core/src/test/scala/org/apache/spark/sql/delta/DeltaSuite.scala

```diff
@@ -597,6 +597,171 @@ class DeltaSuite extends QueryTest
+        .write
+        .format("delta")
+        .partitionBy("part")
+        .mode("append")
+        .save(tempDir.getCanonicalPath)
+
+      Seq(("d", "z")).toDF("value", "part")
+        .write
+        .format("delta")
+        .partitionBy("part")
+        .mode("overwrite")
+        .option("partitionOverwriteMode", "dynamic")
+        .save(tempDir.getCanonicalPath)
+      checkDatasetUnorderly(data.select("value", "part").as[(String, String)],
+        ("a", "x"), ("b", "y"), ("c", "x"), ("d", "z"))
+    }
+  }
+
+  test("batch write: append, dynamic partition overwrite multiple partition columns") {
+    withTempDir { tempDir =>
+      def data: DataFrame = spark.read.format("delta").load(tempDir.toString)
+
+      Seq(("a", "x", 1), ("b", "y", 2), ("c", "x", 3)).toDF("part1", "part2", "value")
+        .write
+        .format("delta")
+        .partitionBy("part1", "part2")
+        .mode("append")
+        .save(tempDir.getCanonicalPath)
+
+      Seq(("a", "x", 4), ("d", "x", 5)).toDF("part1", "part2", "value")
+        .write
+        .format("delta")
+        .partitionBy("part1", "part2")
+        .mode("overwrite")
+        .option("partitionOverwriteMode", "dynamic")
+        .save(tempDir.getCanonicalPath)
+      checkDatasetUnorderly(data.select("part1", "part2", "value").as[(String, String, Int)],
+        ("a", "x", 4), ("b", "y", 2), ("c", "x", 3), ("d", "x", 5))
+    }
+  }
+
+  test("batch write: append, dynamic partition overwrite without partitionBy") {
+    withTempDir { tempDir =>
+      def data: DataFrame = spark.read.format("delta").load(tempDir.toString)
+
+      Seq(1, 2, 3).toDF
+        .withColumn("part", $"value" % 2)
+        .write
+        .format("delta")
+        .partitionBy("part")
+        .mode("append")
+        .save(tempDir.getCanonicalPath)
+
+      Seq(1, 5).toDF
+        .withColumn("part", $"value" % 2)
+        .write
+        .format("delta")
+        .mode("overwrite")
+        .option("partitionOverwriteMode", "dynamic")
+        .save(tempDir.getCanonicalPath)
+      checkDatasetUnorderly(data.select("value").as[Int], 1, 2, 5)
+    }
+  }
+
+  test("batch write: append, dynamic partition overwrite with replaceWhere") {
+    withTempDir { tempDir =>
+      def data: DataFrame = spark.read.format("delta").load(tempDir.toString)
+
+      Seq((1, "x"), (2, "y"), (3, "z")).toDF("value", "part2")
+        .withColumn("part1", $"value" % 2)
+        .write
+        .format("delta")
+        .partitionBy("part1", "part2")
+        .mode("append")
+        .save(tempDir.getCanonicalPath)
+
+      val e = intercept[AnalysisException] {
+        Seq((3, "x"), (5, "x")).toDF("value", "part2")
+          .withColumn("part1", $"value" % 2)
+          .write
+          .format("delta")
+          .partitionBy("part1", "part2")
+          .mode("overwrite")
+          .option(DeltaOptions.REPLACE_WHERE_OPTION, "part1 = 1")
+          .option("partitionOverwriteMode", "dynamic")
```

Comment:
why is one option using string, the other constant. might as well both use the raw string in the same way user would. 

And there should be a test for DPO set through conf as well if we are failing both.

Reply:
Should we also test that the dataframewriter option takes precedence?

i.e. `spark.sql.sources.partitionOverwriteMode=Dynamic` and `.option("partitionOverwriteMode", "static")`

Reply:
> why is one option using string, the other constant. might as well both use the raw string in the same way user would.
> 
> And there should be a test for DPO set through conf as well if we are failing in that scenario as well.

i noticed all other tests use constants. so i was being inconsistent in using raw string. change all tests to use `DeltaOptions.PARTITION_OVERWRITE_MODE_OPTION`

Reply:
> We should also probably have a test checking we support the sparkConf setting

added test that uses sql conf setting, and a test that uses sql conf setting and then overrides it with option

Reply:
Is  that the behavior with Parquet in Apache Spark? option overrides conf?

Reply:
and I think you may have missed this one 

> And there should be a test for DPO set through conf as well if we are failing in that scenario as well.
I meant a test where DPO is set through conf but replaceWhere is set in option. that should fail as of now.

Reply:
> Is that the behavior with Parquet in Apache Spark? option overrides conf?

yes. we rely on this all the time... we have the `partitionOverwriteMode` set to `dynamic` by default in sql conf, and on a per write basis if one needs it to be static for a file source the fix is to add `.option("partitionOverwriteMode", "static")`

Reply:
> and I think you may have missed this one
> 
> > And there should be a test for DPO set through conf as well if we are failing in that scenario as well.
> > I meant a test where DPO is set through conf but replaceWhere is set in option. that should fail as of now.

i now added test for this